// @lucid-agents/ap2/dist/index.js
// src/runtime.ts
function createAP2Runtime(config) {
  if (!config) {
    return void 0;
  }
  return {
    config
  };
}

// src/types.ts
var AP2_EXTENSION_URI = "https://github.com/google-agentic-commerce/ap2/tree/v0.1";

// src/manifest.ts
function createAgentCardWithAP2(card, ap2Config) {
  if (!ap2Config.roles?.length) {
    return card;
  }
  const [firstRole, ...restRoles] = ap2Config.roles;
  if (!firstRole) {
    throw new Error("AP2 roles must contain at least one entry");
  }
  const roles = [firstRole, ...restRoles];
  const extension = {
    uri: AP2_EXTENSION_URI,
    description: ap2Config.description ?? "Agent Payments Protocol (AP2)",
    required: ap2Config.required ?? roles.includes("merchant"),
    params: { roles }
  };
  const existing = card.capabilities?.extensions ?? [];
  const withoutAp2 = existing.filter(
    (ext) => !("uri" in ext && ext.uri === AP2_EXTENSION_URI)
  );
  const capabilities = {
    ...card.capabilities,
    extensions: [...withoutAp2, extension]
  };
  return {
    ...card,
    capabilities
  };
}

// src/extension.ts
function ap2(options) {
  let ap2Runtime;
  return {
    name: "ap2",
    build(ctx) {
      if (options) {
        ap2Runtime = createAP2Runtime(options);
        return { ap2: ap2Runtime };
      }
      return { ap2: void 0 };
    },
    onManifestBuild(card, runtime) {
      if (ap2Runtime?.config) {
        return createAgentCardWithAP2(card, ap2Runtime.config);
      }
      return card;
    }
  };
}

export { AP2_EXTENSION_URI, ap2, createAP2Runtime, createAgentCardWithAP2 };
