// x402-hono/dist/cjs/session-token.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/session-token.ts
var session_token_exports = {};
__export(session_token_exports, {
  POST: () => POST
});
module.exports = __toCommonJS(session_token_exports);
var import_auth = require("@coinbase/cdp-sdk/auth");
async function POST(c) {
  try {
    const apiKeyId = process.env.CDP_API_KEY_ID;
    const apiKeySecret = process.env.CDP_API_KEY_SECRET;
    if (!apiKeyId || !apiKeySecret) {
      console.error("Missing CDP API credentials");
      return c.json(
        { error: "Server configuration error: Missing CDP API credentials" },
        { status: 500 }
      );
    }
    const body = await c.req.json();
    const { addresses, assets } = body;
    if (!addresses || !Array.isArray(addresses) || addresses.length === 0) {
      return c.json(
        { error: "addresses is required and must be a non-empty array" },
        { status: 400 }
      );
    }
    const jwt = await (0, import_auth.generateJwt)({
      apiKeyId,
      apiKeySecret,
      requestMethod: "POST",
      requestHost: "api.developer.coinbase.com",
      requestPath: "/onramp/v1/token"
    });
    const tokenRequestPayload = {
      addresses: addresses.map((addr) => ({
        address: addr.address,
        blockchains: addr.blockchains || ["base"]
      })),
      ...assets && { assets }
    };
    const response = await fetch("https://api.developer.coinbase.com/onramp/v1/token", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${jwt}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify(tokenRequestPayload)
    });
    if (!response.ok) {
      const errorText = await response.text();
      console.error("Failed to generate session token:", response.status, errorText);
      return c.json(
        { error: "Failed to generate session token" },
        { status: response.status }
      );
    }
    const data = await response.json();
    return c.json(data);
  } catch (error) {
    console.error("Error generating session token:", error);
    return c.json({ error: "Internal server error" }, { status: 500 });
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  POST
});
